exonum/src/blockchain/block.rs

```diff
@@ -12,33 +12,127 @@
+}
+
+impl Block {
+    /// Create new `Block`.
+    pub fn new(
         proposer_id: ValidatorId,
-        /// Height of the block, which is also the number of this particular
-        /// block in the blockchain.
         height: Height,
-        /// Number of transactions in this block.
         tx_count: u32,
-        /// Hash link to the previous block in the blockchain.
         prev_hash: &Hash,
-        /// Root hash of the Merkle tree of transactions in this block.
         tx_hash: &Hash,
-        /// Hash of the blockchain state after applying transactions in the block.
         state_hash: &Hash,
+    ) -> Self {
+        Self {
+            proposer_id,
+            height,
+            tx_count,
+            prev_hash: *prev_hash,
+            tx_hash: *tx_hash,
+            state_hash: *state_hash,
+        }
+    }
+    /// Identifier of the leader node which has proposed the block.
+    pub fn proposer_id(&self) -> ValidatorId {
+        self.proposer_id
+    }
+    /// Height of the block, which is also the number of this particular
+    /// block in the blockchain.
+    pub fn height(&self) -> Height {
+        self.height
+    }
+    /// Number of transactions in this block.
+    pub fn tx_count(&self) -> u32 {
+        self.tx_count
+    }
+    /// Hash link to the previous block in the blockchain.
+    pub fn prev_hash(&self) -> &Hash {
+        &self.prev_hash
+    }
+    /// Root hash of the Merkle tree of transactions in this block.
+    pub fn tx_hash(&self) -> &Hash {
+        &self.tx_hash
+    }
+    /// Hash of the blockchain state after applying transactions in the block.
+    pub fn state_hash(&self) -> &Hash {
+        &self.state_hash
+    }
+}
+
+impl ToProtobuf for Block {
+    type ProtoStruct = protobuf::Block;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        let mut msg = Self::ProtoStruct::new();
+        msg.set_proposer_id(self.proposer_id.to_pb_field());
+        msg.set_height(self.height.to_pb_field());
+        msg.set_tx_count(self.tx_count.to_pb_field());
+        msg.set_prev_hash(self.prev_hash.to_pb_field());
+        msg.set_tx_hash(self.tx_hash.to_pb_field());
+        msg.set_state_hash(self.state_hash.to_pb_field());
+        msg
+    }
+
+    fn from_pb(mut pb: Self::ProtoStruct) -> Result<Self, ()> {
+        Ok(Self {
+            proposer_id: ProtobufValue::from_pb_field(pb.get_proposer_id())?,
+            height: ProtobufValue::from_pb_field(pb.get_height())?,
+            tx_count: ProtobufValue::from_pb_field(pb.get_tx_count())?,
+            prev_hash: ProtobufValue::from_pb_field(pb.take_prev_hash())?,
+            tx_hash: ProtobufValue::from_pb_field(pb.take_tx_hash())?,
+            state_hash: ProtobufValue::from_pb_field(pb.take_state_hash())?,
+        })
+    }
+}
+
+impl CryptoHash for Block {
+    fn hash(&self) -> Hash {
+        let v = self.to_pb().write_to_bytes().unwrap();
+        crypto::hash(&v)
+    }
+}
+
+impl StorageValue for Block {
```

Comment:
> Or will it break something for types like Hash?

Exactly!

But I think this implementation will be generated explicitly for `ToProtobuf` type but for other reason: it is impossible to write generic implementation of `StorageValue` for trait `ToProtobuf` and `protobuf::Message`(so that user could use both approaches with explicit conversions or using protobuf messages directly)